[SR-237][build-script] Migrate Ninja build to Python

rintaro opened the pull request:
#### What's in this pull request?

More work on [[SR-237] Merge `build-script-impl` into `build-script`](https://bugs.swift.org/browse/SR-237).

Migrate Ninja build to Python.
Also, some functional changes:

* In OS X, use user configurable `toolchain.cxx` instead of `xcrun --find clang++` to build Ninja.
* When the system installed Ninja is not found, try to build Ninja if `--foundation` is requested as well as `cmake_generator == "Ninja"`.
* To build Foundation, even without `--build-ninja`, use `build-script` detected Ninja path instead of just `ninja`,  because the system installed Ninja can be named `ninja-build`.
* Added debug trace output for commands used in `build-script`
  * Clean build directory `rm -rf {build_root}`
  * Create build directory `mkdir -p {build_root}`
  * Archive symbols package: `tar -c -z -f ...`
* Possible regression:
  * [Existence check of products source directories](https://github.com/apple/swift/blob/7218c41e6012b9cbbb247454485436d8f6cb9531/utils/build-script-impl#L850-L878) has been postponed until after Ninja was built.
  * [Report stdlib targets](https://github.com/apple/swift/blob/7218c41e6012b9cbbb247454485436d8f6cb9531/utils/build-script-impl#L1047-L1056), e.g. `Building the standard library for: ...`, after Ninja was built.
    (Depends #2497)

Please tell me these changes are unacceptable.

Migrated impl args:
`--build-ninja`
`--darwin-deployment-version-{osx,ios,tvos,watchos}`

Removed impl args:
`--build-ninja`: Not used in `build-script-impl` anymore

Added impl args:
`--ninja-bin`: To propagate just built `ninja` path to `build-script-impl`

--

To make it testable, cherry-picked (and enhanced) dry-runnable shell interface from #2241.


#### Resolved bug number: (Partially [SR-237](https://bugs.swift.org/browse/SR-237))

* * * *

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

gribozavr commented on utils/build-script:
```diff
@@ -1200,8 +1230,45 @@ details of the setups of other systems or automated environments.""")
+
+    # Clean build direcotry if requested.
+    if args.clean:
+        shell.rmtree(workspace.build_root)
```
The `if args.show_sdks:` and `if args.clean:` statements seem to be interrupting the flow here.  Is there an ordering dependency?  If not, could you reorder this code so that first we show SDKs, then clean, then configure ninja and build it?

gribozavr commented on utils/swift_build_support/swift_build_support/shell.py:
```diff
@@ -0,0 +1,98 @@
+    return pipes.quote(str(arg))
+
+
+def _coerce_dry_run(dry_run_in):
```
Could you rename `dry_run_in` => `dry_run_override`?

gribozavr commented on utils/swift_build_support/swift_build_support/xcrun.py:
```diff
@@ -43,3 +43,19 @@ def find(tool, sdk=None, toolchain=None):
+    command = ['xcrun', '--sdk', sdk, '--show-sdk-path']
+    try:
+        out = subprocess.check_output(command,
+                                      stderr=subprocess.PIPE)
```
Does `stderr=subprocess.PIPE` mean that the stderr would not be printed to the terminal?  Any particular reason to do this?  I'm afraid that in case `xcrun` fails in CI (but still exits with 0 exit code) we'd be left with no clues left in the build log.

rintaro replied:
I just copy/pasted from `find` without deep thought. No reason to discard stderr here will fix.

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@rintaro Thank you very much for the change and tests!

gribozavr commented:
@swift-ci Please test

gribozavr commented:
@rintaro LGTM, just a few small comments.

rintaro commented:
Fixed. And added missing `xcrun.sdk_path` test case.
Omitting `stderr=subprocess.PIPE` mess up `python -m unittest discover` output, but it's harmless.

gribozavr commented:
@swift-ci Please test
